pontoon/administration/management/commands/sync_projects.py

```diff
@@ -0,0 +1,327 @@
+        self.update_resources(db_project, vcs_project)
+
+        vcs_entities = self.get_vcs_entities(vcs_project)
+        db_entities = self.get_db_entities(db_project)
+
+        # Find all possible entity keys.
+        entity_keys = set().union(db_entities.keys(), vcs_entities.keys())
+
+        changeset = ChangeSet(db_project, vcs_project)
+        for key in entity_keys:
+            db_entity = db_entities.get(key, None)
+            vcs_entity = vcs_entities.get(key, None)
+            self.handle_entity(changeset, db_project, key, db_entity, vcs_entity)
+
+        # Apply the changeset to the files and then commit them.
+        changeset.execute()
+        self.commit_changes(db_project)
+
+        # Once we've successfully committed, update the last_synced date.
+        db_project.last_synced = timezone.now()
+        db_project.save()
+
+    def handle_entity(self, changeset, db_project, key, db_entity, vcs_entity):
+        key = db_entity.string if db_entity else vcs_entity.string
+        self.debug(u'Handling entity: `{0}`'.format(key))
+
+        # If the VCS no longer has the entity, we remove it from
+        # Pontoon.
+        if vcs_entity is None:
+            if db_entity is None:
+                # This should never happen. What? Hard abort.
+                raise CommandError('No entities found for key {0}'.format(key))
+            else:
+                # VCS no longer has the entity, remove it from Pontoon.
+                changeset.remove_db_entity(db_entity.pk)
+                self.debug('  Remove from Pontoon')
+        elif db_entity is None:
+            # New VCS entities are added to Pontoon.
+            changeset.create_db_entity(vcs_entity)
+            self.debug('  Add to Pontoon')
+        else:
+            for locale in db_project.locales.all():
+                if db_entity.has_changed(locale.code):
+                    # Pontoon changes overwrite whatever VCS has.
+                    changeset.update_vcs_entity(locale.code, db_entity, vcs_entity)
+                    self.debug('  Update VCS')
+                elif vcs_entity.has_translation_for(locale.code):
+                    # If Pontoon has nothing or has not changed, and the VCS
+                    # still has the entity, update Pontoon with whatever may
+                    # have changed.
+                    changeset.update_db_entity(locale.code, db_entity, vcs_entity)
+                    self.debug('  Update Pontoon')
+
+    def update_resources(self, db_project, vcs_project):
+        relative_paths = vcs_project.resources.keys()
+        db_project.resource_set.exclude(path__in=relative_paths).delete()
+
+        for relative_path, vcs_resource in vcs_project.resources.items():
+            resource, created = db_project.resource_set.get_or_create(path=relative_path)
+            resource.format = Resource.get_path_format(relative_path)
+            resource.entity_count = len(vcs_resource.entities)
+            resource.save()
```

Comment:
Because we're using the manager from `db_project.resource_set`, the project is implicitly set to `db_project`, and the path is set by `get_or_create`.